torch/fx/experimental/symbolic_shapes.py

```diff
@@ -186,79 +176,217 @@ def magic_impl(self, other):
+        symbols = list(expr.free_symbols)
+        new_shape_env = {
+            k: sympy.Symbol(f"shape_{idx}", positive=True, integer=True) + 1
+            for idx, k in enumerate(symbols)
+        }
+        new_expr = expr.xreplace(new_shape_env)
+        new_expr = sympy.expand(new_expr)
+        if len(list(new_expr.free_symbols)) == 0:
+            return new_expr
+        return None
 
-        self.guards.append((expr, concrete_val))
+    @_lru_cache
+    def replace(self, expr: "sympy.Expr") -> "sympy.Expr":
+        replacements = {s: self._find(s) for s in expr.free_symbols}
+        return sympy.expand(expr.xreplace(replacements))
+
+    @_lru_cache
+    def _update_divisible(self):
+        new_divisible = {}
+        for k in self.divisible:
+            res = self.replace(k)
+            if len(res.free_symbols) > 0:
+                new_divisible[k] = 0
+
+        self.divisible = new_divisible
+
+    @_lru_cache
+    def simplify(self, expr: "sympy.Expr") -> "sympy.Expr":
+        expr = self.replace(expr)
+        if len(expr.atoms(sympy.Mod)) > 0:
+            self._update_divisible()
+            expr = expr.xreplace(self.divisible)
+            expr = sympy.expand(expr)
+        return expr
+
+    @lru_cache(256)
+    def size_hint(self, expr: "sympy.Expr"):
+        result_expr = sympy.expand(expr).xreplace(self.var_to_val)
+        assert len(result_expr.free_symbols) == 0, "Size hint has variables we don't have underlying values for"
+        return result_expr
+
+    def _find(self, a):
+        """
+        Implements a DSU to find the variable that represents a
+        TODO: Improve this to handle non-identity transitive replacements
+        """
+        acopy = a
+        while a in self.replacements:
+            a = self.replacements[a]
+        while acopy in self.replacements:
+            self.replacements[acopy], acopy = a, self.replacements[acopy]
+        return a
+
+    def _maybe_guard_eq(self, expr: "sympy.Eq") -> None:
+        """
+        Evaluates the result of an eq call. If true, uses information to
+        simplify shapes (i.e. a == b or a % 5 == 0)
+        """
+        concrete_bool = bool(self.size_hint(expr))
+        if not concrete_bool:
+            return
+        free = list(expr.free_symbols)
+
+        assert len(free) > 0, "The expression should not be static by this point"
+        # In case of really gnarly expression, we don't blow up
+        if len(free) <= 4:
+            free = sorted(free, key=lambda x: (self.size_hint(x), x.name), reverse=True)  # type: ignore[attr-defined]
+            lhs = expr.lhs
+            rhs = expr.rhs
+            try:
+                solutions = sympy.solveset(lhs - rhs, free[0], domain=sympy.S.Integers)
+                if not solutions.is_finite_set:
+                    if len(expr.atoms(sympy.Mod)) == 1:
+                        mod_expr = tuple(expr.atoms(sympy.Mod))[0]
```

Comment:
🤔 why tuple atoms to immediately retrieve first element

Reply:
atoms is a set.

Reply:
You can do this with a tuple unpack

```
mod_expr, = expr.atoms(sympy.Mod)
```